### PR TITLE
Add tests to checkout

### LINKED ISSUE #507: Checkout tests: test cases on checkout object
We need to check following scenarios
- [x] set shipping address and check if checkout object is modified
- [x] set shipping method and check if checkout object is modified
- [x] set billing address and check if checkout object is modified
- [x] set email address and check if checkout object is modified
- [x] pass the same address twice and check method `is_shipping_same_as_billing`
- [x] pass two different addresses and check method `is_shipping_same_as_billing`
- [x] mock cart object (with partition) and check test method `get_subtotal`
- [x] mock cart object (with partition) and check test method `get_total`
- [x] mock cart object (with partition) and check test method `get_total_shipping`

### PR DESCRIPTION
**Pull request by maciekjablonski**
Fixes #507 

Work in progress

### CODE REVIEW

**Comment by maciekjablonski** [comment]
@patrys  review please

**Comment by patrys** [comment]
It would be helpful to always pass `spec=` when instantiating a Mock. It's both safer (it only allows existing attributes) and easier to read. I can't really tell from reading this code what it does.

**Comment by maciekjablonski** [comment]
@patrys review please

File: tests/test_checkout.py [hunk 1]
```diff
@@ -10,9 +11,56 @@
...
+
+@pytest.fixture
+def cart_with_partition_factory():
+    def generate_cart(items):
+        partitions = []
+        for item in items:
+            is_shipping_required = item.get('is_shipping_required', True)
+            money_function_mock = Mock(return_value=Price(item['cost'],
+                                                          currency=settings.DEFAULT_CURRENCY))
+            item = Mock(spec=['get_total', 'get_price_per_item'],
```
Comment (patrys) [inline tests/test_checkout.py]: Pass a class to `spec`.

**Comment by patrys** [comment]
Can we use lambdas instead of mocks for functions and methods?

**Comment by maciekjablonski** [comment]
@patrys rebased

**Comment by patrys** [comment]
See my comment above

**Comment by artursmet** [comment]
@maciekjablonski Please rebase with master

**Comment by patrys** [comment]
The specs passed to Mock should be proper classes. Passing a list of strings means we now have two places where an API is declared.

**Comment by maciekjablonski** [comment]
@patrys Mock uses model spec now

File: tests/test_checkout.py [hunk 2]
```diff
@@ -1,18 +1,66 @@
...
+    return Checkout(Mock(), AnonymousUser(), 'tracking_code')
+
+
+@pytest.fixture
+def cart_with_partition_factory(cart, monkeypatch, product_in_stock):
+    def generate_cart(items):
+        partitions = []
+        for item in items:
+            is_shipping_required = item.get('is_shipping_required', True)
+            money_function_mock = Mock(return_value=Price(item['cost'],
```
Comment (patrys) [inline tests/test_checkout.py]: Please break the line after the opening parenthesis. Keeping these aligned is a major PITA and there's no obvious way to increase indentation if you open another parenthesis.

File: tests/test_checkout.py [hunk 3]
```diff
@@ -1,18 +1,66 @@
...
+                                                          currency=settings.DEFAULT_CURRENCY))
+            item = Mock(spec=CartLine,
+                        get_total=money_function_mock,
+                        get_price_per_item=money_function_mock)
+
+            partition = Mock(spec=ProductGroup,
```
Comment (patrys) [inline tests/test_checkout.py]: Why do we need to mock the partitioner? Why do we need to mock the cart?

**Comment by maciekjablonski** [comment]
@patrys i changed code, please review again

File: tests/test_checkout.py [hunk 4]
```diff
@@ -193,3 +178,95 @@ def test_checkout_discount(request_cart, sale, product_in_stock):
...
+    assert anonymous_checkout.storage['email'] == 'test@example.com'
+
+
+@pytest.mark.parametrize('items, subtotal', [
+    ([{'cost': 10}, {'cost': 20, 'is_shipping_required': False}], 30),
+    ([{'cost': 10}], 10),
+])
+def test_get_subtotal(anonymous_checkout, cart_with_partition_factory, items, subtotal):
+    """
+    mock cart object (with partition) and check test method get_subtotal
```
Comment (krzysztofwolski) [inline tests/test_checkout.py]: Format docstrings and change to:
> Test checkout subtotal cost calculation for cart with multiple groups

File: tests/test_checkout.py [hunk 5]
```diff
@@ -193,3 +178,95 @@ def test_checkout_discount(request_cart, sale, product_in_stock):
...
+    assert anonymous_checkout.get_subtotal() == Price(subtotal, currency=settings.DEFAULT_CURRENCY)
+
+
+@pytest.mark.parametrize('items, total', [
+    ([{'cost': 10}, {'cost': 20, 'is_shipping_required': False}], 30),
+    ([{'cost': 10}], 10),
+])
+def test_get_total(anonymous_checkout, cart_with_partition_factory, items, total):
+    """
+    mock cart object (with partition) and check test method get_total
```
Comment (krzysztofwolski) [inline tests/test_checkout.py]: Format docstrings and change to:
> Test checkout total cost calculation for cart with multiple groups

File: tests/test_checkout.py [hunk 6]
```diff
@@ -193,3 +178,95 @@ def test_checkout_discount(request_cart, sale, product_in_stock):
...
+    assert anonymous_checkout.get_total() == Price(total, currency=settings.DEFAULT_CURRENCY)
+
+
+@pytest.mark.parametrize('items, total_shipping', [
+    ([{'cost': 10}, {'cost': 20, 'is_shipping_required': False}], 5),
+    ([{'cost': 10}], 5),
+])
+def test_get_total_shipping(anonymous_checkout, cart_with_partition_factory,
+                            shipping_method_factory, items, total_shipping):
+    """
```
Comment (krzysztofwolski) [inline tests/test_checkout.py]: Format docstrings and change to:
> Test checkout total cost calculation for cart with multiple groups and shipping cost

File: tests/test_discounts.py [hunk 7]
```diff
@@ -102,12 +102,12 @@ def test_value_voucher_checkout_discount_not_applicable(settings):
 
 
 @pytest.mark.parametrize(
-    'shipping_cost, shipping_country_code, discount_value, discount_type, apply_to, expected_value', [  # noqa
+    'shipping_cost, shipping_country_code, discount_value, discount_type, apply_to, expected_value', [  # noqa  pylint: disable=line-too-long
```
Comment (krzysztofwolski) [inline tests/test_discounts.py]: Does pylint ignores # noqa ?

Comment (maciekjablonski) [inline tests/test_discounts.py]: noqa is flake8 command

**Pull request closed without merging**